### PR TITLE
[CK_TILE][FMHA] Integrate FAv2 & FAv3 (WIP) in the single fmha_fwd() API

### PR DESCRIPTION
**Pull request by poyenc**
## Proposed changes

This PR merges the two APIs—fmha_fwd() and fmha_fwd_v3()—into a single unified interface. The same script, fmha_fwd.py, is now used to generate two underlying implementation functions: fmha_fwd_v2() and fmha_fwd_v3(). The public API fmha_fwd() conditionally dispatches to fmha_fwd_v3(), although the fmha_fwd_v3() path is temporarily disabled for now (the full implementation is not ready to merge due to compiler issues).

In addition, I redesigned the code-generation logic to allow users to generate multiple dispatcher functions and organize pipelines using appropriate filters.

## Checklist

Please put an `x` into the boxes that apply. You can also fill these out after creating the PR. If you're not sure, please don't hesitate to ask.

- [x] I have added tests relevant to the introduced functionality, and the unit tests are passing locally
- [ ] I have added the test to REGRESSION_TESTS list defined at the top of CMakeLists.txt in tests/CMakeLists.txt, **IF** the test takes more than 30 seconds to run.
- [ ] I have added inline documentation which enables the maintainers with understanding the motivation
- [ ] I have removed the stale documentation which is no longer relevant after this pull request
- [ ] (If this change is user-facing) I have added release notes which provide the end users with a brief summary of the improvement from this pull request
- [x] I have run `clang-format` on all changed files
- [x] Any dependent changes have been merged

## Discussion

If this is a relatively large or complex change, feel free to start a discussion by explaining why you chose the solution you did and what alternatives you considered

### CODE REVIEW

**Review by asleepzzz — DISMISSED** [review]

File: example/ck_tile/01_fmha/codegen/ops/fmha_fwd.py [hunk 1]
```diff
@@ -634,16 +717,179 @@ def api_trait(self) -> FmhaFwdApiTrait:
...
+    rules: Iterable[CompatibilityRule],
+    *,
+    short_circuit: bool = True,
+) -> bool:
+    if short_circuit:
+        for rule in rules:
+            if not rule(problem_ctx, kernel_ctx):
+                return False
+        return True
+    return all(rule(problem_ctx, kernel_ctx) for rule in rules)
```
Comment (qianfengz) [inline example/ck_tile/01_fmha/codegen/ops/fmha_fwd.py]: Is there any real difference between the `short_circuit` path and `all(rule(...))` path ?

File: example/ck_tile/01_fmha/codegen/ops/fmha_fwd.py [hunk 2]
```diff
@@ -634,16 +717,179 @@ def api_trait(self) -> FmhaFwdApiTrait:
...
+                return False
+
+            # only qr_async_trload_v3 use km0=256 & 8-warps
+            is_v3_dedicated_tile = (
+                kernel_ctx.tile.F_bm0 == 256
+                and (kernel_ctx.tile.F_rm0 * kernel_ctx.tile.F_rn0 * kernel_ctx.tile.F_rk0) == 8
+                and (kernel_ctx.tile.F_rm1 * kernel_ctx.tile.F_rn1 * kernel_ctx.tile.F_rk1) == 8
+            )  # fmt: skip
+            is_v3_pipeline = kernel_ctx.pipeline.tag == "qr_async_trload_v3"
+            return is_v3_dedicated_tile == is_v3_pipeline
```
Comment (qianfengz) [inline example/ck_tile/01_fmha/codegen/ops/fmha_fwd.py:873]: This is not a rule to restrict the `problem_ctx` and `kernel_ctx`,   can the rule be solved by adding restrictions when constructing the `kernel_ctx` space ?

File: example/ck_tile/01_fmha/codegen/ops/fmha_fwd.py [hunk 3]
```diff
@@ -634,16 +717,179 @@ def api_trait(self) -> FmhaFwdApiTrait:
...
+            if problem_ctx.dtype != "fp32":
+                # TODO: update if >=gfx11 archs get qr_async and qr_async_trload support
+                if kernel_ctx.pipeline.tag in cls._AVAILABLE_PIPELINES and (
+                    (
+                        (problem_ctx.hdim, problem_ctx.hdim_v) == (128, 128)
+                        and kernel_ctx.tile.F_bn0 != 128
+                    )
+                    or (
+                        (problem_ctx.hdim, problem_ctx.hdim_v) != (128, 128)
+                        and kernel_ctx.tile.F_bm0 != 128
```
Comment (qianfengz) [inline example/ck_tile/01_fmha/codegen/ops/fmha_fwd.py:830]: This restriction makes no sense!  (bm0=64 should be able to be used with other hdim other 128)

File: example/ck_tile/01_fmha/codegen/ops/fmha_fwd.py [hunk 4]
```diff
@@ -634,16 +717,179 @@ def api_trait(self) -> FmhaFwdApiTrait:
...
+
+        def check_hdim(problem_ctx: ProblemContext, kernel_ctx: KernelContext) -> bool:
+            # NOTE: this is used to speedup deepseek prefill case, we don't gen training
+            if (problem_ctx.hdim, problem_ctx.hdim_v) == (192, 128):
+                if (
+                    kernel_ctx.pipeline.F_bias != "no"
+                    or kernel_ctx.pipeline.F_dropout == "t"
+                ):
+                    False
+            return True
```
Comment (qianfengz) [inline example/ck_tile/01_fmha/codegen/ops/fmha_fwd.py:792]: This rule makes no sense!   Whether a pipeline can use bias or dropout should have nothing to do with hdim sizes

File: example/ck_tile/01_fmha/codegen/ops/fmha_fwd.py [hunk 5]
```diff
@@ -634,16 +717,179 @@ def api_trait(self) -> FmhaFwdApiTrait:
...
+            problem_ctx: ProblemContext, kernel_ctx: KernelContext
+        ) -> bool:
+            # logits_soft_cap is only allowed if no bias
+            if not (
+                (
+                    kernel_ctx.pipeline.F_logits == "t"
+                    and kernel_ctx.pipeline.F_bias == "no"
+                )
+                or kernel_ctx.pipeline.F_logits == "f"
+            ):
```
Comment (qianfengz) [inline example/ck_tile/01_fmha/codegen/ops/fmha_fwd.py:804]: Can this rule be solved inside the `kernel_ctx` space since it does not involve `problem_ctx` ?

File: example/ck_tile/01_fmha/codegen/ops/fmha_fwd.py [hunk 6]
```diff
@@ -68,50 +69,49 @@
...
+using fmha_kernel = {F_kernel}<fmha_pipeline, fmha_epilogue>;
 
-using trait_{F_idx} = fmha_fwd_traits_<{F_hdim}, {F_dtype}, {F_mode},{F_bm0}, {F_bn0}, {F_bk0}, {F_bn1}, {F_bk1}, {F_bk0max}, {F_vlayout},
-                        {F_pipeline_enum}, {F_logits}, fmha_mask_{F_idx}, {F_bias}, {F_lse}, {F_dropout}, {F_squant}, {F_spad}, {F_skpad}, {F_dpad}, {F_dvpad}, {F_trload}, {F_skip}>;
+using trait = fmha_fwd_traits_<{F_hdim}, {F_dtype}, {F_mode},{F_bm0}, {F_bn0}, {F_bk0}, {F_bn1}, {F_bk1}, {F_bk0max}, {F_vlayout},
+                        {F_pipeline_enum}, {F_logits}, fmha_mask, {F_bias}, {F_lse}, {F_dropout}, {F_squant}, {F_spad}, {F_skpad}, {F_dpad}, {F_dvpad}, {F_trload}, {F_skip}>;
 
 template<>
-float fmha_fwd_<trait_{F_idx}, {F_arch.tag}>(const ck_tile::stream_config& s, fmha_fwd_args a)
+float fmha_fwd_<trait, {F_arch.tag}>(const ck_tile::stream_config& s, fmha_fwd_args a)
```
Comment (qianfengz) [inline example/ck_tile/01_fmha/codegen/ops/fmha_fwd.py:119]: Do we really need `trait` as a template of  `fmha_fwd_`

Comment (poyenc) [reply example/ck_tile/01_fmha/codegen/ops/fmha_fwd.py:119]: Yes. By current design, we use the trait as a instance key to differentiate each template instantiations

File: example/ck_tile/01_fmha/codegen/ops/fmha_fwd.py (hunk 4 above)
Comment (poyenc) [reply example/ck_tile/01_fmha/codegen/ops/fmha_fwd.py:792]: This is a pre-existing check logics for the qr_async_trload pipeline. If you'd remove it, we can create another PR to the purpose.

File: example/ck_tile/01_fmha/codegen/ops/fmha_fwd.py (hunk 5 above)
Comment (poyenc) [reply example/ck_tile/01_fmha/codegen/ops/fmha_fwd.py:804]: That would be another type of check that only consider the kernel_ctx attributes. We can seperate it if we later encounter more checks like this

File: example/ck_tile/01_fmha/codegen/ops/fmha_fwd.py (hunk 3 above)
Comment (poyenc) [reply example/ck_tile/01_fmha/codegen/ops/fmha_fwd.py:830]: This is a pre-existing check logics. If you'd remove it, we can create another PR to the purpose.

File: example/ck_tile/01_fmha/codegen/ops/fmha_fwd.py (hunk 1 above)
Comment (poyenc) [reply example/ck_tile/01_fmha/codegen/ops/fmha_fwd.py]: There should have no difference because I didn't create a new list as function argument. I'll remove the short_circuit path

**Comment by poyenc** [comment]
Need to resolve the conflicts.

**Review by asleepzzz — APPROVED** [review]

**Review by qianfengz — APPROVED** [review]
